actors/builtin/miner/miner_actor.go

```diff
@@ -1363,24 +1363,33 @@ func (a Actor) ReportConsensusFault(rt Runtime, params *ReportConsensusFaultPara
 	}
 
 	// Reward reporter with a share of the miner's current balance.
-	slasherReward := RewardForConsensusSlashReport(faultAge, rt.CurrentBalance())
-	_, code := rt.Send(reporter, builtin.MethodSend, nil, slasherReward)
-	builtin.RequireSuccess(rt, code, "failed to reward reporter")
-
 	var st State
 	rt.State().Readonly(&st)
+	unlockedBalance := st.GetUnlockedBalance(rt.CurrentBalance())
+	slasherReward := RewardForConsensusSlashReport(faultAge, rt.CurrentBalance())
+	if slasherReward.GreaterThan(unlockedBalance) {
+		rt.Abortf(exitcode.ErrInsufficientFunds, "insufficient unlocked balance to pay reward")
```

Comment:
This is surprising. It seems to reward the miner (by reverting the slash) and penalize the slasher (by not paying them) for the miner being under-collateralized. Is this right? Can you add a comment explaining why?

Reply:
Glad you pointed it out.  Before the change this was deducting a fixed portion of the actor balance without any checks on whether that amount was less than the unlocked funds so it was possible to violate the balance invariants and fail in an assertion (which I stumbled upon while testing).  This state is just an explicit version of what we had before so now the uncanny behavior is visible and we will be motivated to design it away.

The simplest alternative is to take the reward from the miner actor in the normal penalization order.  I didn't want to do that in this PR because there is something potentially bad about allowing the reward to come from locked funds -- if miners slash themselves they could unvest  their rewards early.  This might warrant coming up with a penalization order that does not draw from locked funds.

I am planning to file an issue about this to gather input from red team and crypto econ.

Reply:
This will be resolved in this PR now that I understand the path forward